### DISCUSSION TITLE
Help with Texture

### DISCUSSION THREAD

**Discussion by mjehrhart** [emilk/egui#1191] [Q&A]
Hey,

I've been trying this for a bit and I've had no luck at all.

`

    fn update(&mut self, ctx: &egui::CtxRef, frame: &epi::Frame) {

      egui::SidePanel::right("right_panel").show(ctx, |ui| {
            let mut widget_size = egui::Vec2::new(200.0, 100.0);

            let image = load_image(include_bytes!("/Users/me/tmp/bucket/jj.png")).unwrap();

            let ( (x,y), widget_size) = image_background();
            println!("x::{}", x);
            println!("y::{}", y);
            let size = (x,y);

            let tex_allocator = frame.tex_allocator().unwrap();
            let texture_id = tex_allocator.alloc_srgba_premultiplied(size, widget_size);
 
        });
 }

fn image_background() -> ((usize, usize), Vec<Color32>) {
    let image_data_bkg = include_bytes!("/Users/matthew/tmp/bucket/jj.png");
    let image_bkg = image::load_from_memory(image_data_bkg).expect("Failed to load image");
    let image_buffer_bkg = image_bkg.to_rgba8();
    let size_bkg = (5200 as usize, 2500 as usize);
    let pixels_bkg = image_buffer_bkg.into_vec();
    let pixels_bkg: Vec<_> = pixels_bkg
        .chunks_exact(4)
        .map(|p| egui::Color32::from_rgba_unmultiplied(p[0], p[1], p[2], p[3]))
        .collect();
 
    (size_bkg, pixels_bkg)
}

with this code here, the error I get is on "tex_allocator".

`error[E0599]: no method named `tex_allocator` found for reference `&eframe::epi::Frame` in the current scope
   --> src/main.rs:218:39
    |
218 |             let tex_allocator = frame.tex_allocator().unwrap();
    |                                       ^^^^^^^^^^^^^ method not found in `&eframe::epi::Frame`

`

I noticed in someone's example that frame used a lifetime [ frame: &epi::Frame<'_' ] but the compiler complains about that as well. I could use some help here.

**Comment by mjehrhart** [comment]
Any thoughts on this anyone?

the error message is:

```no method named \`tex_allocator\` found for struct \`eframe::epi::Frame\` in the current scope
method not found in \`eframe::epi::Frame\`rustcE0599```


here's an updated code segment:

```
fn image_background(frame: epi::Frame) -> ((usize, usize), Vec<Color32>) {
    let image_data_bkg = include_bytes!("/Users/matthew/tmp/bucket/jj.png");
    let image_bkg = image::load_from_memory(image_data_bkg).expect("Failed to load image");
    let image_buffer_bkg = image_bkg.to_rgba8();
    let size_bkg = (5200 as usize, 2500 as usize);
    let pixels_bkg = image_buffer_bkg.into_vec();
    let pixels_bkg: Vec<_> = pixels_bkg
        .chunks_exact(4)
        .map(|p| egui::Color32::from_rgba_unmultiplied(p[0], p[1], p[2], p[3]))
        .collect(); 
    let texture_bkg = frame
                .tex_allocator()
                .alloc_srgba_premultiplied(200., &500);
 
    (size_bkg, pixels_bkg)
}
```

**Reply by emilk** [reply]
This was changed recently;  see the changelog for `eframe`: https://github.com/emilk/egui/blob/master/eframe/CHANGELOG.md

Also, read the docs: https://docs.rs/epi/latest/epi/struct.Frame.html

**Comment by mjehrhart** [comment]
Thanks for the response.  I see what you are referring to. 

 So I tried the example at https://github.com/emilk/egui/blob/master/eframe/examples/image.rs but ran into the same issues as the user who commented on bug 1110 - https://github.com/emilk/egui/pull/1110.

My update fn requires CtxRef and not Context, maybe there's an issue there. I am using egui = "0.16.1".

**Reply by emilk** [reply]
The latest `master` contains the examples that are compatible with the latest `master` of `egui`.

If you are using egui 0.16, you need to look at the example for egui 0.16: https://github.com/emilk/egui/blob/0.16.0/eframe/examples/image.rs

**Reply by mjehrhart** [reply]
thank you so much, that cleared up so much for me. thanks.
